smack-extensions/src/main/java/org/jivesoftware/smackx/jingle/transports/jingle_s5b/elements/JingleS5BTransport.java

```diff
@@ -0,0 +1,144 @@
+        this.dstAddr = dstAddr;
+        this.mode = mode;
+    }
+
+    public String getStreamId() {
+        return streamId;
+    }
+
+    public String getDestinationAddress() {
+        return dstAddr;
+    }
+
+    public Bytestream.Mode getMode() {
+        return mode == null ? Bytestream.Mode.tcp : mode;
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE_V1;
+    }
+
+    @Override
+    protected void addExtraAttributes(XmlStringBuilder xml) {
+        xml.optAttribute(ATTR_DSTADDR, dstAddr);
+        xml.optAttribute(ATTR_MODE, mode);
+        xml.attribute(ATTR_SID, streamId);
+    }
+
+    public boolean hasCandidate(String candidateId) {
+        return getCandidate(candidateId) != null;
+    }
+
+    public JingleS5BTransportCandidate getCandidate(String candidateId) {
+        for (JingleContentTransportCandidate c : candidates) {
+            JingleS5BTransportCandidate candidate = (JingleS5BTransportCandidate) c;
+            if (candidate.getCandidateId().equals(candidateId)) {
+                return candidate;
+            }
+        }
+        return null;
+    }
+
+    public static Builder getBuilder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        private String streamId;
+        private String dstAddr;
+        private Bytestream.Mode mode;
+        private ArrayList<JingleContentTransportCandidate> candidates = new ArrayList<>();
+        private ArrayList<JingleContentTransportInfo> infos = new ArrayList<>();
+
+        public Builder setStreamId(String sid) {
+            this.streamId = sid;
+            return this;
+        }
+
+        public Builder setDestinationAddress(String dstAddr) {
+            this.dstAddr = dstAddr;
+            return this;
+        }
+
+        public Builder setMode(Bytestream.Mode mode) {
+            this.mode = mode;
+            return this;
+        }
+
+        public Builder addTransportCandidate(JingleS5BTransportCandidate candidate) {
+            this.candidates.add(candidate);
+            return this;
+        }
+
+        public Builder addTransportInfo(JingleContentTransportInfo info) {
+            this.infos.add(info);
+            return this;
+        }
+
+        public Builder setCandidateUsed(String candidateId) {
+            return addTransportInfo(JingleS5BTransportInfo.CandidateUsed(candidateId));
+        }
+
+        public Builder setCandidateActivated(String candidateId) {
+            return addTransportInfo(JingleS5BTransportInfo.CandidateActivated(candidateId));
+        }
+
+        public Builder setCandidateError() {
+            return addTransportInfo(JingleS5BTransportInfo.CandidateError());
+        }
+
+        public Builder setProxyError() {
+            return addTransportInfo(JingleS5BTransportInfo.ProxyError());
+        }
+
+        public JingleS5BTransport build() {
+            StringUtils.requireNotNullOrEmpty(streamId, "sid MUST be neither null, nor empty.");
```

Comment:
verification is usually performed as late as possible, which means in the constructor. not in the build() method